@varungup90 

after a stress-test, we found that if add routerCtx.Delete() in process, an close channel error will occur, which may means we call it more than once

```
func (s *Server) Process(srv extProcPb.ExternalProcessor_ProcessServer) error {
	st := &processState{
		ctx:       srv.Context(),
		requestID: uuid.New().String(),
	}
	defer func() {
		if st.routerCtx != nil {
			st.routerCtx.Delete()
		}
	}()
```

```
I0526 06:49:48.844965       1 gateway_req_body.go:152] "request_start" request_id="16483eb2-86be-476c-b4fd-43f1c439fdd4" request_path="/v1/chat/completions" model="mock" stream=true routing_strategy="prefix-cache" target_pod="chat-mock-ai-ff5ff55f6-hh64q" target_pod_ip="10.233.174.243:8000" outstanding_requests=14 routing_time_taken="850.074µs"
I0526 06:49:48.845165       1 gateway.go:134] "processing request" requestID="0357227b-0330-49b6-8b52-177b2e7aa2c6"
W0526 06:49:48.845229       1 pod.go:356] requestID: 0357227b-0330-49b6-8b52-177b2e7aa2c6, pod:  is missing port identifier label: model.aibrix.ai/port, hence default to port: 8000
I0526 06:49:48.845262       1 gateway_req_body.go:152] "request_start" request_id="5f0e8f56-25a6-4bf3-9225-2f66e84f11ff" request_path="/v1/chat/completions" model="mock" stream=true routing_strategy="prefix-cache" target_pod="" target_pod_ip=":8000" outstanding_requests=0 routing_time_taken="560ns"
panic: close of closed channel

goroutine 5357 [running]:
github.com/vllm-project/aibrix/pkg/types.(*RoutingContext).SetTargetPod(0x4001ccc9c0, 0x1ecf130?)
	github.com/vllm-project/aibrix/pkg/types/router_context.go:218 +0x78
github.com/vllm-project/aibrix/pkg/types.(*RoutingContext).Delete(0x4001ccc9c0)
	github.com/vllm-project/aibrix/pkg/types/router_context.go:154 +0x24
github.com/vllm-project/aibrix/pkg/plugins/gateway.(*Server).Process.func1()
	github.com/vllm-project/aibrix/pkg/plugins/gateway/gateway.go:130 +0x28
github.com/vllm-project/aibrix/pkg/plugins/gateway.(*Server).Process(0x40004760e0, {0x2c7c670, 0x4007ee8d10})
	github.com/vllm-project/aibrix/pkg/plugins/gateway/gateway.go:140 +0x40c
github.com/envoyproxy/go-control-plane/envoy/service/ext_proc/v3._ExternalProcessor_Process_Handler({0x252bee0?, 0x40004760e0}, {0x2c79380, 0x400662a5a0})
	github.com/envoyproxy/go-control-plane@v0.12.0/envoy/service/ext_proc/v3/external_processor.pb.go:1847 +0xdc
google.golang.org/grpc.(*Server).processStreamingRPC(0x40002e2200, {0x2c70878, 0x4004c587e0}, {0x2c80520, 0x4001cf4a80}, 0x4004c8a120, 0x4001b98c30, 0x41c4a40, 0x0)
	google.golang.org/grpc@v1.65.0/server.go:1673 +0xe60
google.golang.org/grpc.(*Server).handleStream(0x40002e2200, {0x2c80520, 0x4001cf4a80}, 0x4004c8a120)
	google.golang.org/grpc@v1.65.0/server.go:1794 +0xaec
google.golang.org/grpc.(*Server).serveStreams.func2.1()
	google.golang.org/grpc@v1.65.0/server.go:1029 +0x8c
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 917
	google.golang.org/grpc@v1.65.0/server.go:1040 +0x13c
```

and we found  a delete() called in HandleResponseBody and HandleResponseHeaders, so i decided to revert last change( adding defer routerCtx.Delete())

```
	defer func() {
		// Wrapped in a function to delay the evaluation of parameters. Using complete to make sure DoneRequestTrace only call once for a request.
		if !hasCompleted && complete {
			s.cache.DoneRequestTrace(routerCtx, requestID, model, promptTokens, completionTokens, traceTerm)
			if routerCtx != nil {
				routerCtx.Delete()
			}
		}
	}()
```

--- a/pkg/plugins/gateway/gateway.go
+++ b/pkg/plugins/gateway/gateway.go
@@ -125,11 +125,6 @@ func (s *Server) Process(srv extProcPb.ExternalProcessor_ProcessServer) error {
 		ctx:       srv.Context(),
 		requestID: uuid.New().String(),
 	}
-	defer func() {
-		if st.routerCtx != nil {
-			st.routerCtx.Delete()
-		}
-	}()
 
 	klog.InfoS("processing request", "requestID", st.requestID)
 	labels := map[string]string{"pod_name": podName}
@@ -333,6 +328,11 @@ func (s *Server) sendProcessingResponse(srv extProcPb.ExternalProcessor_ProcessS
 		klog.ErrorS(err, "gateway fail to send response to envoy-proxy", "requestID", st.requestID)
 		s.emitMetricsCounterHelper(metrics.GatewayRequestModelFailTotal, st.model, "send_envoy_proxy", "499")
 		s.cache.DoneRequestCount(st.routerCtx, st.requestID, st.model, st.traceTerm)
+		// Manually delete routerCtx on error paths;
+		// HandleResponseBody() and HandleResponseHeaders() will handle it on the happy path.
+		if st.routerCtx != nil {
+			st.routerCtx.Delete()
+		}
 
 		if errors.Is(err, context.Canceled) || strings.Contains(err.Error(), "EOF") {
 			klog.Warning("Stream already closed by client", "requestID", st.requestID)

